File: src/chipschallenge/tickbehaviors/TankBehavior.java
package chipschallenge.tickbehaviors;

import chipschallenge.Block;
import chipschallenge.BlockContainerFullException;
import chipschallenge.Move;
import chipschallenge.buttonbehaviors.ButtonBehavior;

/**
 *
 * @author patrik
 */
public class TankBehavior extends ButtonBehavior implements BlockTickBehavior {

    private boolean isMoving = false;

    @Override
    public void tick(Block caller) throws BlockContainerFullException {
        if (isMoving) {
            if (!caller.move(caller.getFacing())) {
                isMoving = false;
            }
        }
    }

    public void buttonDown(Block listener, Block button) {
        listener.setFacing(Move.reverse(listener.getFacing()));
        isMoving = true;
    }

    public void buttonUp(Block listener, Block button) {
        // Do nothing
    }
}

File: ChipsChallenge.Shared/Tickbehaviors/TankBehavior.cs
namespace ChipsChallenge.Shared.Tickbehaviors
{
    using Buttonbehaviors;

    public class TankBehavior : IButtonBehavior, IBlockTickBehavior
    {
        private bool isMoving = true;

        public virtual void Tick(Block caller)
        {
            if (isMoving)
            {
                if (!caller.Move(caller.Facing))
                {
                    isMoving = false;
                }
            }
        }

        public virtual void ButtonDown(Block listener, Block button)
        {
            listener.Facing = Move.Reverse(listener.Facing);
            isMoving = true;
        }

        public virtual void ButtonUp(Block listener, Block button)
        {
        }
    }
}